File: dotnet/samples/SimpleOrm.Sample/Repositories/UserRepository.cs
using SimpleOrm.Sample.Models;

namespace SimpleOrm.Sample.Repositories;

/// <summary>
/// Per-entity repository: the generic surface comes from the library's
/// <see cref="Repository{TEntity}"/> (ADR-0016); only entity-specific reads live
/// here — one-line criteria, no per-table SQL.
/// </summary>
public sealed class UserRepository(Db db) : Repository<User>(db)
{
    public Task<User> GetByEmailAsync(string email, CancellationToken ct)
        => Query().Where(Criteria.Eq(nameof(User.Email), email)).SingleAsync(ct);

    public Task<IReadOnlyList<User>> GetByIdsAsync(IReadOnlyList<long> ids, CancellationToken ct)
        => Query().Where(Criteria.In(nameof(User.Id), ids)).OrderBy(nameof(User.Id)).ToListAsync(ct);
}

File: php/samples/SimpleOrm.Sample/src/Repositories/UserRepository.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Sample\Repositories;

use SimpleOrm\Query\Criteria;
use SimpleOrm\Sample\Models\Tables\User;
use SimpleOrm\Session\Db;

/**
 * Per-entity repository: the generic surface comes from {@see Repository}
 * (ADR-0016); only entity-specific reads live here, as one-line criteria, no
 * per-table SQL.
 *
 * @extends Repository<User>
 */
final class UserRepository extends Repository
{
    public function __construct(Db $db)
    {
        parent::__construct($db, User::class);
    }

    public function getByEmail(string $email): User
    {
        /** @var User $user */
        $user = $this->query()->where(Criteria::eq('email', $email))->single();

        return $user;
    }

    /**
     * @param list<int> $ids
     * @return list<User>
     */
    public function getByIds(array $ids): array
    {
        /** @var list<User> $users */
        $users = $this->query()->where(Criteria::in('id', $ids))->orderBy('id')->toList();

        return $users;
    }
}
